Helper/MerchantMetrics.php

```diff
@@ -0,0 +1,343 @@
+        $rootPath  =  $directory->getRoot();
+        return $rootPath . '/bolt_metrics.json';
+    }
+
+    /**
+     * Attempts to open a file and if it cannot open in 5 seconds it will return null
+     *
+     * @return Stream
+     */
+    public function waitForFile(){
+        $count = 0;
+        $timeoutMsecs = 10; //number of miliseconds of timeout (10 * 500) so 5 seconds
+
+        if ($this->metricsFile == null) {
+            $this->metricsFile = $this->setFile();
+        }
+
+        $workingFile = fopen($this->metricsFile, "a+");
+
+        // logic for properly grabbing file and locking it
+        while (!$this->lockFile($workingFile)) {
+            if ($count++ < $timeoutMsecs) {
+                usleep(500);
+            } else {
+                return null;
+                break;
+            }
+        }
+        return $workingFile;
+    }
+
+    /**
+     * Add a count metric to the array of metrics being stored
+     *
+     * @param string        $key    name of count metric
+     * @param int           $value  count hit 
+     *
+     * @return void
+     */
+    public function addCountMetric($key, $value)
+    {
+        $data = [
+                'value' => $value,
+                "metric_type" => "count",
+                "timestamp" => $this->getCurrentTime(),
+            ];
+        $this->metrics[$key] = $data;
+    }
+
+     /**
+     * Add a latency metric to the array of metrics being stored
+     *
+     * @param string        $key     name of latency metric
+     * @param int           $value  the total time of the metric
+     *
+     * @return void
+     */
+    public function addLatencyMetric($key, $value)
+    {
+        $data = [
+                'value' => $value,
+                "metric_type" => "latency",
+                "timestamp" => $this->getCurrentTime(),
+            ];
+        $this->metrics[$key] = $data;
+
+    }
+
+    /**
+     * Adds a count and latency for a given event
+     *
+     * @param string        $countKey           name of count metric
+     * @param int           $countValue         the count value of the metric
+     * @param string        $latencyKey         name of latency metric
+     * @param int           $latencyValue  the total time of the metric
+     *
+     * @return void
+     */
+    public function processMetrics($countKey, $countValue, $latencyKey, $latencyValue)
+    {
+
+
+        $test = $this->configHelper->shouldCaptureMerchantMetrics();
+        // checks if flag is set for Merchant Metrics if not ignore
+        if ($test) {
+            $this->addCountMetric($countKey, $countValue);
+            $this->addLatencyMetric($latencyKey, $latencyValue);
+
+            if ($this->metricsFile == null) {
+                $this->metricsFile = $this->setFile();
+            }
+            $workingFile = $this->waitForFile();
+            if ($workingFile) {
+                file_put_contents($this->metricsFile, json_encode($this->metrics), FILE_APPEND);
```

Comment:
I don't think using member variable metrics here is a good idea. suppose there are two thread executing processMetrics, you could be writing this content into file twice? you use file system to lock things but member variable live independent of file system.

I think `addCountMetric` and `addLatencyMetric` should be renamed to `formatCountMetric` which just returns php object, and in this processMetrics method we maintain list of metrics to write.

Reply:
I added this logic and split the dependency of count and latency. Is it worth keeping the combined one as well to reduce the amount of file writes we are doing? That is the one benefit there.

Reply:
I start to wonder if you even need count metrics. you can just use count of latency metrics if you always report both count and metrics at the same location. I suggest you start to build sample dashboard in datadog and talk with merchant exp team to figure out what metrics are valuable. 

For scope of this PR, I'm ok with creating utility function to report both count and metrics but as I said above reporting only latency should probably just work fine in that case.